HPCC-24026 Ensure k8s jobnames are unique

jakesmith opened the pull request:
Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-13741, label: hpcc-24026-prevent-k8s-jobname-clash starts now.
The reason of this test is: Forced to re-test.
Commit ID: 49d424bdfc10e4343934146b1b8eac8899eb6304
Estimated completion time is ~0.5 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-201.ca-central-1.compute.internal

jakesmith commented:
@richardkchapman - please review.

Re. whitespaces (trimmed trailing), this is being caused by an option I have set in VSCode, to prevent trailing white spaces, which automatically strips any trailing whitespaces from end of lines when a file is saved.
It's convenient and avoids them being introduced and clears up the existing erroneous spaces on a per-file basis. Let me know if you okay with that in this/other commits, e.g. by reviewing in github by turning off white spaces from the diff.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-201.ca-central-1.compute.internal
Sha: 49d424bdfc10e4343934146b1b8eac8899eb6304
Build: success
Milestone:Install hpccsystems-platform-community_7.8.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |54 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |25 sec  (00:00:25) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |17 sec  (00:00:17) |
|test (hthor) |874 |874 |0 |422 sec  (00:07:02) |
|test (thor) |812 |812 |0 |749 sec  (00:12:29) |
|test (roxie) |953 |953 |0 |373 sec  (00:06:13) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  643 sec (00:10:43) |  118 sec (00:01:58) |   22 sec (00:00:22) |   18 sec (00:00:18) | 1867 sec (00:31:07) |   18 sec (00:00:18) | 2695 sec (00:44:55) |

HPCCSmoketest commented:
Process of PR-13741, label: hpcc-24026-prevent-k8s-jobname-clash starts now.
The reason of this test is: Did not build yet.
Commit ID: 49d424bdfc10e4343934146b1b8eac8899eb6304
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: 49d424bdfc10e4343934146b1b8eac8899eb6304
Build: success
Milestone:Install hpccsystems-platform-community_7.8.13-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |53 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |44 sec  (00:00:44) |
|setup (roxie) |11 |11 |0 |19 sec  (00:00:19) |
|test (hthor) |874 |874 |0 |682 sec  (00:11:22) |
|test (thor) |812 |812 |0 |1004 sec  (00:16:44) |
|test (roxie) |953 |953 |0 |709 sec  (00:11:49) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   12 sec (00:00:12) |  932 sec (00:15:32) |  123 sec (00:02:03) |   23 sec (00:00:23) |   17 sec (00:00:17) | 2747 sec (00:45:47) |   18 sec (00:00:18) | 3872 sec (01:04:32) |

richardkchapman commented:
I can't see the jira so there may be more commentary there, but it's not clear to me why a particular wu/graph would be executing more than once?

jakesmith commented:
The ECL construct "RECOVERY" can cause it.
It can re-run the same graph (and hence subgraph) again after failure.
I'll add the particular regression suite query name to the JIRA for reference that caused this issue to be hit.

jakesmith commented:
Well I would if JIRA wasn't down. The test query is : issue22196.ecl

richardkchapman commented on roxie/ccd/ccdcontext.cpp:
```diff
@@ -2317,11 +2317,13 @@ class CRoxieContextBase : implements IRoxieSlaveContext, implements ICodeContext
+        {
+            CDateTime dt;
+            dt.setNow();
+            VStringBuffer job("%s-%s-%" I64F "u", wuid.str(), graphName, dt.getHash(0));
```
Feels like the logic should be in runK8sjob to me. I'm also wondering whether (since this is a very rare/unusual situation) we would be better adding a suffix only if job creation fails? Depends how much we care about job names being easy to type/decode...

jakesmith committed:
HPCC-24026 Replace existing jobs if they exist.

In some circumstances we resubmit a job with the same name,
after the previous incarnation has failed, e.g. due to RECOVERY
option. If any existing job exists it should be replaced.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

HPCCSmoketest commented:
Process of PR-13741, label: hpcc-24026-prevent-k8s-jobname-clash starts now.
The reason of this test is: Forced to re-test.
Commit ID: a237bdb12a5ddf84b0461f7f44f9008d510b5e29
Estimated completion time is ~0.5 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-240.ca-central-1.compute.internal

jakesmith commented:
@richardkchapman - please re-review/merge.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-240.ca-central-1.compute.internal
Sha: a237bdb12a5ddf84b0461f7f44f9008d510b5e29
Build: success
Milestone:Install hpccsystems-platform-community_7.8.15-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |53 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |7 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |29 sec  (00:00:29) |
|setup (thor) |11 |11 |0 |43 sec  (00:00:43) |
|setup (roxie) |11 |11 |0 |20 sec  (00:00:20) |
|test (hthor) |874 |874 |0 |367 sec  (00:06:07) |
|test (thor) |812 |812 |0 |808 sec  (00:13:28) |
|test (roxie) |953 |953 |0 |379 sec  (00:06:19) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  643 sec (00:10:43) |  129 sec (00:02:09) |   22 sec (00:00:22) |   19 sec (00:00:19) | 1882 sec (00:31:22) |   18 sec (00:00:18) | 2722 sec (00:45:22) |

HPCCSmoketest commented:
Process of PR-13741, label: hpcc-24026-prevent-k8s-jobname-clash starts now.
The reason of this test is: New commit.
Commit ID: a237bdb12a5ddf84b0461f7f44f9008d510b5e29
Estimated completion time is ~0.9 hour(s)
centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.8.2003 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-116.ca-central-1.compute.internal
Sha: a237bdb12a5ddf84b0461f7f44f9008d510b5e29
Build: success
Milestone:Install hpccsystems-platform-community_7.8.15-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |120 |120 |0 |0 |0 |54 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |2 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |8 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |28 sec  (00:00:28) |
|setup (thor) |11 |11 |0 |46 sec  (00:00:46) |
|setup (roxie) |11 |11 |0 |25 sec  (00:00:25) |
|test (hthor) |874 |874 |0 |626 sec  (00:10:26) |
|test (thor) |812 |812 |0 |1005 sec  (00:16:45) |
|test (roxie) |953 |953 |0 |701 sec  (00:11:41) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  919 sec (00:15:19) |  123 sec (00:02:03) |   26 sec (00:00:26) |   17 sec (00:00:17) | 2670 sec (00:44:30) |   18 sec (00:00:18) | 3782 sec (01:03:02) |
